### PR TITLE
✨ Add GitHub folder URL support for ERD viewer

### PR DESCRIPTION
**Pull request by FunamaYukina**
## Issue

- resolve: https://github.com/route06/liam-internal/issues/5723

## Why is this change needed?

This adds comprehensive GitHub folder URL support to the ERD viewer, allowing users to paste GitHub repository folder URLs (like `https://github.com/user/repo/tree/main/schemas`) directly into the ERD viewer. The system automatically detects, fetches, and combines all schema files from the folder to generate an ERD diagram.

## Summary

- **GitHub folder URL detection and parsing** - Added `isGitHubFolderUrl()` and `parseGitHubFolderUrl()` utilities
- **GitHub Contents API integration** - Implemented recursive folder traversal using GitHub's REST API
- **Intelligent schema file detection** - Smart filtering based on file extensions with blacklist for utility files
- **Automatic format detection** - Detects schema format (Prisma, SQL, Drizzle, etc.) from file extensions with priority ordering
- **File combination with boundary comments** - Combines multiple files with clear separation for better readability
- **Comprehensive error handling** - Uses Result pattern with neverthrow for robust error management
- **Full test coverage** - 12 test cases covering URL validation, API mocking, error scenarios, and edge cases
- **Code quality compliance** - Reduced cognitive complexity from 23 to ~8 through helper function extraction

⚠️ The performance was not a concern, so no cache control has been implemented.

## Testing
[sadmann7/skateshop](https://github.com/sadmann7/skateshop/tree/main/src/db/schema) (15 files, 16 tables)

https://liam-app-git-feature-github-folder-url-support-liambx.vercel.app/erd/p/github.com/sadmann7/skateshop/tree/main/src/db/schema

Since it searches for files recursively, you can specify the parent folder and it will still display them.
https://liam-app-git-feature-github-folder-url-support-liambx.vercel.app/erd/p/github.com/sadmann7/skateshop/tree/main/src/db

ss 3914

[noodle-run/noodle](https://github.com/noodle-run/noodle/tree/main/src/db/schema) (3 files, 2 tables)

https://liam-app-git-feature-github-folder-url-support-liambx.vercel.app/erd/p/github.com/noodle-run/noodle/tree/main/src/db/schema

ss 3913

🤖 Generated with [Claude Code](https://claude.ai/code)

## Summary by CodeRabbit

- New Features
  - Import ERD schemas directly from GitHub folder URLs; relevant files are combined and format is auto-detected.
  - Added format detection for Drizzle in .ts/.js and tbls in .json files.
- Bug Fixes
  - More consistent error states when loading schemas from URLs, with safeguards for timeouts, file sizes, and nested folders.
- Tests
  - Expanded coverage for GitHub URL handling and format detection.
- Chores
  - Updated dependencies to improve reliability and error handling.

### CODE REVIEW

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.test.ts [hunk 1]
```diff
@@ -0,0 +1,388 @@
+import { afterEach, describe, expect, it, vi } from 'vitest'
+import {
+  fetchSchemaFromGitHubFolder,
+  isGitHubFolderUrl,
+  parseGitHubFolderUrl,
+} from './githubUrlHandler'
+
+// Mock fetch globally
+global.fetch = vi.fn()
```
Comment (MH4GF) [inline frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.test.ts]: nitpack: For mocking network responses, I recommend using the MSW tool since it's already installed, but it's up to you.

https://github.com/liam-hq/liam/blob/974f9a479fbe5556a2bfb59c29cf0b94b7603e9d/frontend/internal-packages/agent/src/checkpoint/SupabaseCheckpointSaver.test.ts#L22

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.test.ts [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+        'https://github.com/user/repo/tree/dev/db',
+      ]
+
+      for (const url of fileUrls) {
+        expect(isGitHubFolderUrl(url)).toBe(false)
+      }
+      for (const url of folderUrls) {
+        expect(isGitHubFolderUrl(url)).toBe(true)
+      }
+    })
```
Comment (MH4GF) [inline frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.test.ts]: What is the difference between this test and the two tests above? If they are duplicates, let's delete them.

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    // SchemaRB non-schema files
+    'seeds.rb',
+    'application_record.rb',
+    // Drizzle non-schema files
+    'migrate.ts',
+    // Documentation
+    'readme.md',
+    'readme.sql',
+    'readme.prisma',
+  ]
```
Comment (MH4GF) [inline frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts:94]: It's tricky to cover all these cases, so it's a bit of a headache. I think we should leave it as is for now. We'll just have to improve it while monitoring the errors that come up in Sentry.

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+  if (excludedFilenames.includes(lowerFilename)) {
+    return false
+  }
+
+  return true
+}
+
+export const parseGitHubFolderUrl = (url: string): GitHubRepoInfo | null => {
+  try {
```
Comment (MH4GF) [inline frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts]: Do not use try-catch; use neverthrow instead.

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+    return { owner, repo, branch, path }
+  } catch {
+    return null
+  }
+}
+
+const fetchGitHubFolderContents = async (
+  repoInfo: GitHubRepoInfo,
+): Promise<Result<GitHubContentItem[], Error>> => {
+  const apiUrl = `https://api.github.com/repos/${repoInfo.owner}/${repoInfo.repo}/contents/${repoInfo.path}?ref=${repoInfo.branch}`
```
Comment (MH4GF) [inline frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts]: You should use Oktokit. You should implement it in `@liam-hq/github`.

https://github.com/liam-hq/liam/blob/974f9a479fbe5556a2bfb59c29cf0b94b7603e9d/frontend/internal-packages/github/src/api.server.ts

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+    return 'drizzle'
+  }
+  if (extensions.some((ext) => ext === '.sql')) {
+    return 'postgres'
+  }
+  if (extensions.some((ext) => ext === '.json')) {
+    return 'tbls'
+  }
+  return null
+}
```
Comment (MH4GF) [inline frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts:383]: Please use this function.


https://github.com/liam-hq/liam/blob/974f9a479fbe5556a2bfb59c29cf0b94b7603e9d/frontend/packages/schema/src/parser/supportedFormat/detectFormat.ts

File: frontend/apps/app/app/erd/p/[...slug]/page.tsx [hunk 7]
```diff
@@ -38,6 +42,149 @@ const resolveContentUrl = (url: string): string | undefined => {
...
+  return format
+}
+
+// Helper function to render error view
+function renderErrorView(
+  blankSchema: {
+    tables: Record<string, never>
+    enums: Record<string, never>
+    extensions: Record<string, never>
+  },
```
Comment (MH4GF) [inline frontend/apps/app/app/erd/p/[...slug]/page.tsx]: If the schema is definitely empty, you should construct the empty schema within the function itself rather than accepting it as an argument.

**Comment by MH4GF** [comment]
@coderabbit This PR adds a process to recursively read GitHub directories. Please carefully review the security risks this introduces.
Since it only parses strings and converts them into objects of a specific format, we believe there is likely no issue.

**Comment by MH4GF** [comment]
@FunamaYukina I believe CodeRabbit's point is valid. Please consider the limits on the number of recursive calls and the number of files.

**Comment by FunamaYukina** [comment]
@MH4GF 
Thank you so much! In addition to revising the review, I'll also add restrictions!

```
const SECURITY_LIMITS = {
  MAX_RECURSION_DEPTH: 5,
  MAX_FILES_PER_FOLDER: 50,
  MAX_TOTAL_FILES: 100
}
```

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.test.ts (hunk 1 above)
Comment (FunamaYukina) [reply frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.test.ts]: Thank you for your guidance.  I've changed to using the msw tool!

https://github.com/liam-hq/liam/pull/3607/commits/5a928e6bc4ed0cdefef2ce2c903a4471fc2df8a3

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.test.ts (hunk 2 above)
Comment (FunamaYukina) [reply frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.test.ts]: This was a duplicate test. I have deleted it🙏.
https://github.com/liam-hq/liam/pull/3607/commits/18ee9382e81519a214461ee3f6e8f9e52cc4b52f

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts (hunk 3 above)
Comment (FunamaYukina) [reply frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts:94]: You're absolutely right. It's unlikely there's a README file. I've removed the unnecessary file specification.

https://github.com/liam-hq/liam/pull/3607/commits/3d262be0acbc8b567b97c1e32e4bdf2033d5d95e

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts (hunk 4 above)
Comment (FunamaYukina) [reply frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts]: I missed that🙏. I've fixed it to use `neverthrow`.
https://github.com/liam-hq/liam/pull/3607/commits/b943ae06c4d2c788abd0500b9079b5ad5c21d9ea

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts (hunk 6 above)
Comment (FunamaYukina) [reply frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts:383]: Thank you. I've modified it to use `detectFormat`.
https://github.com/liam-hq/liam/pull/3607/commits/89b277065d123b41ebcaeaa5996ab5a2178f88c3

File: frontend/apps/app/app/erd/p/[...slug]/page.tsx (hunk 7 above)
Comment (FunamaYukina) [reply frontend/apps/app/app/erd/p/[...slug]/page.tsx]: Thank you for your guidance!

https://github.com/liam-hq/liam/pull/3607/commits/84c14ba63f814597d7efd88b6ffb44a755dbd30a

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts (hunk 5 above)
Comment (FunamaYukina) [reply frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts]: Thank you for your guidance!🙏

https://github.com/liam-hq/liam/pull/3607/commits/b943ae06c4d2c788abd0500b9079b5ad5c21d9ea

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    hash: '',
+    port: '',
+    username: '',
+    password: '',
+    searchParams: new URLSearchParams(),
+    toString: () => url,
+    toJSON: () => url,
+  }
+
+  return ok(urlObj)
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts]: @coderabbitai
Thank you, I've committed below.
https://github.com/liam-hq/liam/pull/3607/commits/7d40e541053cbe8585202fd6db82a95453a093a8

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+    return err(new Error('Failed to download any schema files'))
+  }
+
+  const combinedContent = validContents
+    .map((content, index) => {
+      return `\n// --- File ${index + 1} ---\n${content}`
+    })
+    .join('\n\n')
+
+  return ok(combinedContent)
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts]: @coderabbitai
Thank you, I've committed below.
https://github.com/liam-hq/liam/pull/3607/commits/5ed0ddf072342ad874cf58212c42ea85449857c3

File: frontend/internal-packages/github/src/api.server.ts [hunk 10]
```diff
@@ -194,3 +194,99 @@ export const getOrganizationInfo = async (
...
+    const response = await fetch(url)
+    if (!response.ok) {
+      console.error(`Failed to download file: ${response.statusText}`)
+      return null
+    }
+    return await response.text()
+  } catch (error) {
+    console.error(`Error downloading file from ${url}:`, error)
+    return null
+  }
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/internal-packages/github/src/api.server.ts:294]: @coderabbitai
Thank you, I've committed below.
https://github.com/liam-hq/liam/pull/3607/commits/b5ca715a1c2e0c060d13eeba1c11de4c59b599d6

File: frontend/internal-packages/github/src/api.server.ts [hunk 11]
```diff
@@ -194,3 +194,146 @@ export const getOrganizationInfo = async (
...
+    // Handle abort/timeout errors
+    if (error instanceof Error && error.name === 'AbortError') {
+      console.error(`Request timeout downloading file from ${url}`)
+      return null
+    }
+
+    console.error(`Error downloading file from ${url}:`, error)
+    return null
+  }
+}
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/internal-packages/github/src/api.server.ts:297]: @coderabbitai Since access is limited to GitHub's host, it seems like there shouldn't be any issues, but what do you think?

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+  if (filesInFolder > SECURITY_LIMITS.MAX_FILES_PER_FOLDER) {
+    return err(
+      new Error(
+        `Too many files in folder (${filesInFolder}). Maximum allowed: ${SECURITY_LIMITS.MAX_FILES_PER_FOLDER}`,
+      ),
+    )
+  }
+
+  return ok(undefined)
+}
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts:229]: @coderabbitai
Thank you, I've committed below.
https://github.com/liam-hq/liam/pull/3607/commits/7fb2ef8fa13559b0abfd9fd1dd03d86ba0d8eec8

File: frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+  }
+
+  const combinedContent = validContents
+    .map((content, index) => {
+      return `\n// --- File ${index + 1} ---\n${content}`
+    })
+    .join('\n\n')
+
+  return ok(combinedContent)
+}
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/apps/app/app/erd/p/[...slug]/utils/githubUrlHandler.ts:371]: @coderabbitai
Thank you, I've committed below.

https://github.com/liam-hq/liam/pull/3607/commits/0f48e44d76a082b932236fcc1f16a62d6ab9d064

**Comment by FunamaYukina** [comment]
@MH4GF 

After your review, I made the following changes:

* Fixed all review comments
* Added restrictions with security in mind:

```js
const SECURITY_LIMITS = {
  MAX_RECURSION_DEPTH: 5,
  MAX_FILES_PER_FOLDER: 50,
  MAX_TOTAL_FILES: 100,
  MAX_FILE_SIZE_BYTES: 5 * 1024 * 1024, // 5MB per file
  MAX_DIRS_PER_FOLDER: 50,
  MAX_BRANCH_SEGMENTS: 5
}
```

Additionally, I added a timeout setting (10s) for `downloadFileContent`.

* Addressed CodeRabbit’s feedback:

  * Fixed the missing consideration for GitHub URLs where the branch name includes a slash (e.g., `feature/xxx`).

**Review by MH4GF — APPROVED** [review]
It works 👍🏻 Thanks!!!!
https://liam-app-git-feature-github-folder-url-support-liambx.vercel.app/erd/p/github.com/sadmann7/skateshop/tree/main/src/db/schema

**Comment by FunamaYukina** [comment]
Thank you for your repeated reviews!!🙏
